Improvements and handle Google errors

rgazelot opened the pull request:
Related to https://github.com/Calendart/CalendArt/pull/6

Taluu commented on src/AbstractApi.php:
```diff
@@ -0,0 +1,68 @@
+                throw new Exception\BackendException;
+        }
+    }
+}
```
how about using a trait instead ?

Taluu commented:
Instead of `AbstractApi`, you should use a trait, and the exceptions are quite related to google point of view IMO...

And the `handleErrors` isn't pretty imo, it should be triggered only on cases where the exception code is _not_ what we expect (not a 200 when we should exepct a 200 ? call that method. Not a 201 ? etc, etc)

rgazelot commented:
> the exceptions are quite related to google point of view IMO

Fortunately, we are in the Google adapter...

rgazelot commented:
@Taluu For the second point, maybe handleErrors is a bad name, this is more an `analyzeResponse`

rgazelot commented:
@Taluu Open for review. (I'll rebase and change the name of some commits :p)

Taluu commented on src/CalendarApi.php:
```diff
@@ -33,6 +28,8 @@
  */
 class CalendarApi implements CalendarApiInterface, AclInterface
 {
+    use ResponseHandler;
```
:+1:

rgazelot commented:
@Taluu I think i'm done with this one.

Taluu commented on test/ResponseHandlerTest.php:
```diff
@@ -0,0 +1,71 @@
+            [500, 'Backend Error', Exception\BackendException::class],
+        ];
+    }
+}
```
instead of making a blank mock `Api` class, what you can do is using a mock that use the trait with `$this->getMockForTrait()` : https://phpunit.de/manual/current/en/test-doubles.html#test-doubles.mocking-traits-and-abstract-classes

Taluu commented:
Please keep in mind we're on php 5.4 >, not php 5.5 (so no `class`...).
I should raise the requirement, but this is not the scope of this PR IMO (or 1.x actually...)

Taluu commented:
This looks good though, so just replace the `::class` with the full qualified names and it's good to go

rgazelot commented:
@Taluu done

rgazelot commented:
@Taluu I fixed the php 5.4 stuffs

rgazelot committed:
Create some proper exception depending on Google responses

Based on the documentation
https://developers.google.com/google-apps/calendar/v3/errors

rgazelot committed:
Handle API responses in a trait

rgazelot committed:
Implement the ResponseHandler

Taluu commented:
:tada:
